_posts/2018-08-09-deploying-esxi-through-foreman.md

```diff
@@ -0,0 +1,402 @@
+  ```
+
+  Note: Instead of mounting the ISO, you can also extract the ISO contents to the directory and edit the boot.cfg directly inside that directory. Remember to change the PXELinux template accordingly. I chose this because this approach is a tad cleaner.  
+
+  Now we are ready to deploy this OS on to a host.
+
+**2.5:** Open https://{foreman-url}/hosts/ and edit the host on which you want to deploy ESXi.  
+
+**2.6:** Under the 'Operating System' tab, select the following values:
+
+  Operating System : ESXi 6.7 Build 8169922  
+  Media : ESXi Dummy  
+  Partition table: Kickstart default  
+  PXE Loader : ESXi PXELinux BIOS  
+
+**2.7:** Press 'Resolve' beside Provisioning templates and you should see the ESXi PXELinux and Kickstart templates.
+  <p float='left' align='left'>
+    <img src="/static/images/blog_images/2018-08-16-deploying-esxi-through-foreman/edit_host_legacy_bios.png" width="400" />
+  </p>
+
+**2.8:** Press 'Build' and reboot the host to PXE. If all goes well, you should see ESXi installer load up and perform an automated installation.
+  <p float='left' align='center'>
+    <img src="/static/images/blog_images/2018-08-16-deploying-esxi-through-foreman/esxi_installer_loading.png" width="400" />
+  </p>
+
+Now once installation is complete and the host reboots, it will try to PXEBoot and **should** be presented with the default PXE menu.
+<p float='left' align='center'>
+  <img src="/static/images/blog_images/2018-08-16-deploying-esxi-through-foreman/localboot.png" width="400" />
+</p>
+Upon timeout, the host should attempt a Localboot and boot to ESXi.
+
+However, since the host will boot into the pxelinux.0 v3.86 bootloader, without the supporting files of the same version (menu.c32 etc) the host will not be presented with this PXE menu and will never boot into ESXi.
+
+To solve this we have to manually change the host PXE Loader to **PXELinux BIOS** or **None** in Foreman.  
+This can be quite tedious when we are deploying at scale. Here is where Foreman-Hooks comes to the rescue.
+
+From the [Foreman-Hooks Github repo](https://github.com/theforeman/foreman_hooks#foreman_hooks):  
+
+> Foreman Hooks allows you to trigger scripts and commands on the Foreman server at any point in an object's lifecycle in Foreman. This lets you run scripts when a host is created, or finishes provisioning etc.
+
+### 3: Using Foreman hooks to solve host not able to perform local boot from PXE
+---
+
+Once the host informs Foreman that build is completed, we will run a script that checks if the host's OS is ESXi and then changes the PXELoader to None. For this we need to hook into the before_provision event.
+
+```sh
+cd /usr/share/foreman/config/hooks
+
+mkdir -p host/managed/before_provision
+
+cd host/managed/before_provision
+
+# Symlink the hook_functions file
+ln -s /usr/share/foreman/vendor/ruby/2.3.0/gems/foreman_hooks-0.3.14/examples/bash/hook_functions.sh .
```

Comment:
This path is dependent on the OS and package version. I think that on RH we place it `/usr/share/doc`. Maybe it's better to copy it?

Reply:
Have added a note at the top that warns about this. Otherwise a lot of paths needed to be changed.